autopush_rs/src/util/user_agent.rs

```diff
@@ -0,0 +1,42 @@
+use woothee::parser::{Parser, WootheeResult};
+
+// List of valid user-agent attributes to keep, anything not in this
+// list is considered 'Other'. We log the user-agent on connect always
+// to retain the full string, but for DD more tags are expensive so we
+// limit to these.
+const VALID_UA_BROWSER: &[&str] = &["Chrome", "Firefox", "Safari", "Opera"];
+
+// See dataset.rs in https://github.com/woothee/woothee-rust for the
+// full list (WootheeResult's 'os' field may fall back to its 'name'
+// field). Windows has many values and we only care that its Windows
+const VALID_UA_OS: &[&str] = &["Firefox OS", "Linux", "Mac OSX"];
+
+pub fn parse_user_agent<'a>(
+    parser: &'a Parser,
+    agent: &str,
+) -> (WootheeResult<'a>, &'a str, &'a str) {
+    let wresult = parser.parse(&agent).unwrap_or_else(|| WootheeResult {
+        name: "",
+        category: "",
+        os: "",
+        os_version: "".to_string(),
+        browser_type: "",
+        version: "".to_string(),
+        vendor: "",
+    });
+
+    // Determine a base os/browser for metrics' tags
+    let metrics_os = if wresult.os.starts_with("Windows") {
```

Comment:
Is case a concern? Should we check these against lowercase only?

Reply:
we're comparing against what woothee gives us which is always title case for these (see https://github.com/woothee/woothee-rust/blob/master/src/dataset.rs). note that woothee gives us "Mac OSX" vs "Mac OS X" that the python parser gave

There is a concern here, though -- this could use some tests..

Reply:
ah, ok 👍